### ISSUE TITLE
Definir la documentación para el feature Usuarios

### ISSUE THREAD

**Issue by bordonwork** [rofaccess/CENADE#28]
Se requiere definir la documentación necesaria en donde se especifique:
### Descripción

En esta sección se deberá describir el feature, su utilidad para el negocio, los actores, las vistas necesarias (y su propósito), entre otro detalles que ayuden a comprender el objeto del feature.  
### Flujo

Para cada formulario del feature se requerirá especificar:
- Flujo del formulario: esto incluye el flujo de carga del formulario, así como lo procesos que ocurren por detrás (actualización de registros por ejemplo).
- Flujo alternativo: otro flujo que pueda seguirse.
- Excepciones: casos en los que no se podrá guardar los datos del formulario.
- Mensajes: de éxito, error o advertencias (excluir aquellos de validación). 

**Nota:** con respecto a los flujos y excepciones, éstos deben redactarse desde el punto de vista del formulario, no del proceso del negocio (por lo tanto se excluyen referencias como: el paciente llega y pide un turno).
### Dependencias

Se debe especificar las dependencias con otros features, y explicar en qué consisten dichas dependencias.
### Gemas necesarias

En esta sección se deberá incluir las gemas que se utilizarán para el feature. Se debe especificar para cada vista del feature que gemas se utilizan, la versión de la gema y para que se utiliza dicha gema.
### Prioridad

La prioridad del feature dentro del proyecto en una escala del 1 al 10 (en donde 10 es la prioridad más alta).

**Comment by rofaccess** [comment]
# Feature Usuarios:

## Descripción

En esta sección se puede agregar, modificar y eliminar usuarios. Cualquier usuario con los permisos correpondientes puede acceder a esta sección.
Se ingresa a la interfaz de usuarios, a través del submenu usuarios del menú Configuración.

## Agregar Usuario

### Flujo Principal
1. En la interfaz de usuarios, se muestra por defecto una tabla con la lista de usuarios. En la parte superior derecha de la interfaz, se tiene el botón “Agregar Usuario”.
2. Al clickear en el botón “Agregar Usuario” se muestra el formulario de “Nuevo Usuario”.
3. En el formulario se tiene un input de búsqueda, que permite buscar a un empleado por su número de CI usando ajax. 
4. Se ingresa el número de CI, nombre o apellido y se clickea en el botón “Buscar”.
5. Si el empleado existe y no tiene usuario, se muestran sus datos básicos, como nombres, apellidos, cargo y email.
   (Nota: Los datos básicos son para asegurarse de que se le va a asignar un usuario al empleado correcto)
6. Se habilita la opción guardar del formulario.
7. Se especifica el nombre de usuario, antes de guardar, se debe validar si ya existe o no un usuario con ese nombre.
8. Se especifica el rol que tendrá el usuario, antes de guardar se debe validar que se haya seleccionado un rol.
9. Se clickea en el botón “Guardar” del formulario.
10. En el controlador se especifica un password para el usuario. El nombre de usuario se utilizará como password.
11. Al guardar el usuario, si no hubo ningún problema, se redirecciona a una pantalla que muestra el usuario recien guardado y se muestra el mensaje: “El usuario nombre_usuario ha sido guardado” (Fin).

### Flujo Alternativo

5.1.  Si el empleado existe y tiene usuario  se muestra el mensaje: “El empleado ya tiene un usuario en el sistema”

5.2. Se muestra sus datos básicos, como nombres, apellidos, cargo y email.

5.3. Se muestra su nombre de usuario y su rol actuales.

5.4. No se permite editar nada. El usuario puede especificar otro número de documento (Pasar a 4).

5.1.  Si el empleado existe y tiene un usuario escondido por el borrado oscioso, se muestra sus datos básicos, como nombres, apellidos, cargo y email (Pasar a 6).
(Nota: Para este caso, vamos a obligar al usuario a que especifique un nuevo nombre de usuario y rol, asegurandonos de actualizar en el controlador, el mismo usuario que ya estaba relacionado con el empleado, desabilitando el borrado ocioso. Se verá como un guardado normal).

11 Al guardar el usuario, si hubo algún problema, se redirecciona a la pantalla de agregar usuario y se muestra el mensaje:  “No se ha podido guardar el usuario nombre_usuario”, si es posible, el mensaje deberá especificar el motivo (Pasar a 4).

### Excepciones

5 Si el empleado no existe, se mostrará un mensaje “No encontrado”. 
(Nota: Se tendrá que registrar primero al empleado, en la interfaz de empleados o especificar otro número de documento).

## Editar Usuario

### Flujo Principal
1. En la interfaz de usuarios, se muestra por defecto una tabla con la lista de usuarios, cada registro de la tabla tiene un botón para editarlo.
2. Al clickear en el botón “Editar” se muestra el formulario de “Editar Usuario”.
3. En el formulario se muestran datos básicos, como nombres, apellidos, cargo y email del empleado al que pertenece el usuario.
4. Se muestra el nombre de usuario actual, antes de guardar, se debe validar si ya existe o no un usuario con ese nombre.
5. Se muestra el rol actual del usuario, antes de guardar se debe validar que se haya seleccionado un rol.
6. Se muestra un checkbox con label “Resetear Password”, esto utilizará el nombre de usuario especificado, como password.
7. Se clickea en el botón “Guardar” del formulario.
8. Al guardar el usuario, si no hubo ningún problema, se redirecciona a una pantalla que muestra el usuario recien editado y se muestra el mensaje: “El usuario nombre_usuario ha sido editado” (Fin).

### Flujo Alternativo

8 Al guardar el usuario, si hubo algún problema, se redirecciona a la pantalla de agregar usuario y se muestra el mensaje:  “No se ha podido guardar el usuario nombre_usuario”, si es posible, el mensaje deberá especificar el motivo (Pasar a 4).

## Eliminar Usuario

### Flujo Principal
1. En la interfaz de usuarios, se muestra por defecto una tabla con la lista de usuarios, cada registro de la tabla tiene un botón para eliminarlo, la opción eliminar también se encuentra en el formulario Mostrar Empleado.
2. Al clickear en el botón “Eliminar” se muestra un mensaje de confirmación: “¿Desea eliminar el usuario nombre_usuario?” con las opciones “Si” “No”.
3. La opción “No” cancela el borrado. La opción “Si” borra el usuario, muestra la lista de usuarios, y muestra el mensaje “El usuario nombre_usuario ha sido eliminado”

## Mostrar Usuario

### Descripción
1. En la interfaz de usuarios, se muestra por defecto una tabla con la lista de usuarios, cada registro de la tabla tiene un botón para ver los detalles del registro.
2. Al clickear en el botón “Ver” se muestran los datos del usuario y datos básicos del empleado.
3. En esta pantalla se tienen las opciones “Listar” (Muestra la lista de empleados), “Eliminar” (Permite eliminar el usuario), “Editar” (Muestra el formulario de edición).

## Listar Usuario

### Descripción
1. En la interfaz de usuarios, se muestra por defecto una tabla con la lista de usuarios, cada registro muestra cierto campos concernietes al usuario y al empleado relacionado.
2. Sobre el listado se tiene un campo de busqueda, para buscar sobre cualquier campo de la lista. También se tiene un botón de “Busqueda Avanzada” que desplegará más campos, para filtrar los datos de la lista, de una manera más específica.
3. En esta pantalla se tienen las opciones: “Nuevo Usuario” (Permite agregar un nuevo usuario), “Ver”, “Editar”, “Eliminar”, por cada registro de la lista
4. La lista tiene paginación.

## Dependencias

Ninguna 

## Gemas necesarias
1. Devise 3.0
2. Ransack 1.7.0
3. Kaminari 0-17

## Prioridad

10

**Comment by rofaccess** [comment]
Espero no haberme pasado con los detalles.

**Tiempo Utilizado:** 03h00m

**Comment by KariSanabria** [comment]
@rofaccess creo que esta muy bien detallado, consulta solo por C.I se podrá buscar al empleado?

**Comment by rofaccess** [comment]
Especifique que solo por CI, es que se espera que todos tengan CI, y para no complicar la implementación usando nombres y apellidos también. Ahora si resulta facil y útil usar nombre y apellido cuando lo estemos implementando, para mi que no habrá drama con hacerlo.

**Comment by KariSanabria** [comment]
Creo que podríamos buscar también por Nombre y Apellido.
Creo que este ya podes cerrar!!
@Lesfig alguna sugerencia?

**Comment by Lesfig** [comment]
Opino igual que Kari, también se debería poder buscar por Nombre y Apellido, lo demás creo que ya está todo bien! :)

**Comment by rofaccess** [comment]
ok. ya le agregue lo de nombre y apellido
